Experimental/src/Ranges/Range.cs

```diff
@@ -0,0 +1,713 @@
+                // ?
+
+                if (other.HasEnd)
+                {
+                    //    ?----------------cE
+                    //          oS-----oE
+                    //          oS-------------oE
+
+                    var compare = other.End.CompareTo(current.End);
+                    if (compare <= 0)
+                    {
+                        //    ?-------------===cE
+                        //          oS-----oE
+
+                        if (compare == 0)
+                        {
+                            //    ?------------cE
+                            //          oS-----oE
+
+                            if (current.IncludeEnd && !other.IncludeEnd)
+                            {
+                                var newRange = new Range<TValue>(other.End, other.End,
+                                    RangeOptions.HasStart | RangeOptions.HasEnd
+                                    | RangeOptions.IncludingStart
+                                    | RangeOptions.IncludingEnd
+                                    );
+                                yield return newRange;
+                            }
+                        }
+                        else
+                        {
+                            var newRange = new Range<TValue>(other.End, current.End,
+                                RangeOptions.HasStart | RangeOptions.HasEnd
+                                | (other.IncludeEnd ? RangeOptions.None : RangeOptions.IncludingStart) // ivert flag
+                                | (current.IncludeEnd ? RangeOptions.IncludingEnd : RangeOptions.None)
+                                );
+
+                            if (!newRange.IsEmpty)
+                                yield return newRange;
+                        }
+                    }
+                }
+            }
+            else
+            {
+                // cS---------------
+                // ?
+                if (other.HasEnd)
+                {
+                    var newRange = new Range<TValue>(other.End, default(TValue),
+                        RangeOptions.HasStart
+                        | (other.IncludeEnd ? RangeOptions.None : RangeOptions.IncludingStart) // ivert flag
+                        );
+
+                    if (!newRange.IsEmpty)
+                        yield return newRange;
+                }
+            }
+        }
+
+        public IEnumerable<Range<TValue>> Intersect(Range<TValue> other)
+        {
+            var current = this;
+
+            if (current.IsEmpty || other.IsEmpty)
+            {
+                yield return Empty;
+                yield break;
+            }
+
+            if (other.IsFull)
+            {
+                yield return this;
+                yield break;
+            }
+
+            if (current.IsFull)
+            {
+                yield return other;
+                yield break;
+            }
+
+
+            if (!current.Intersects(other))
```

Comment:
It looks like this check could be omitted. Instead you can just check at the end that the newStart and newEnd give a valid range